### DISCUSSION TITLE
Candidate window doesn't follow QT theme on Linux

### DISCUSSION THREAD

**Discussion by BrianAllred** [google/mozc#796] [General]
## Description
Candidate window doesn't follow the specified QT theme on Linux, which can make the candidates difficult to read and also cause eye strain (if it defaults to a bright color like white).

## Steps to reproduce
1. Set environment variable `QT_QPA_PLATFORMTHEME` to a dark theme.
2. Type using mozc so that the candidate window appears.

## Expected behavior
The candidate window appears using the correct theme.

## Actual behavior
The candidate window appears using an incorrect theme

## Screenshots
Incorrect theming:
![image](https://github.com/google/mozc/assets/5847511/a0738e83-089d-4ff1-be79-5d12cb4fc620)

For example, this is what anthy looks like:
![image](https://github.com/google/mozc/assets/5847511/db364060-970a-4788-bcfa-340af0ef3b58)

## Version or commit-id
mozc-2.29.5111.102-1.2

## Environment
 - OS: openSUSE Tumbleweed using latest Gnome Wayland
 - IMF: ibus
 - Related Applications: all

## Investigations
 - Whether this issue happens on both Ibus and other IMF (e.g. Fcitx, uim).
   + N/A
 - Whether this issue happens on other IMEs (e.g. Anthy, SKK).
   + no
 - What applications this issue happens on (e.g. Chromium, gedit).
   + all
 - What applications this issue does not happen on (e.g. Chromium, gedit).
   + N/A
 - (optional) What versions or commit-ids this issue did not happen on
   (e.g. Mozc-2.28.4960.100+24.11.oss).
   + N/A

**Comment by yukawa** [comment]
Thank you for the report. Supporting dark theme has been our to-do list for long time.

For your case, you can uninstall `ibus-mozc-candidate-window` package from our environment to use IBus' default candidate window, which is what Anthy is using in your screenshot.

The following document also describes how to use IBus' default candidate window rather than Mozc's one.
https://github.com/google/mozc/blob/master/docs/configurations.md#use-the-default-ibus-candidate-window

As for our long term plan, I guess we may want to rely on [the corresponding settings in XDG Desktop Portal](https://github.com/flatpak/xdg-desktop-portal/issues/629) rather than relying on `QT_QPA_PLATFORMTHEME`, which cannot handle dynamic theme update. Here are several examples in other projects.

 * GNOME: https://blogs.gnome.org/alicem/2021/10/04/dark-style-preference/
 * KDE: https://invent.kde.org/plasma/xdg-desktop-portal-kde/-/merge_requests/49
 * Chromium: https://bugs.chromium.org/p/chromium/issues/detail?id=998903
 * Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1734934

I'm going to convert this to discussion as it's a bit unlikely to just use `QT_QPA_PLATFORMTHEME` anyway.

**Comment by yukawa** [comment]
> I'm going to convert this to discussion as it's a bit unlikely to just use `QT_QPA_PLATFORMTHEME` anyway.

Actually, looks like Qt 6.2 and later have already been reading  XDG Desktop Portal settings.
 * https://github.com/qt/qtbase/commit/3c6b61a6e14962969e7e9dc5909cd746e38f8de7
 * https://codereview.qt-project.org/c/qt/qtbase/+/396655

**Comment by pancaek** [comment]
Hello, I am facing a similar issue with mozc on Manjaro (ibus-mozc-2.29.5200.102-1)
When following the solution described in https://github.com/google/mozc/blob/master/docs/configurations.md#use-the-default-ibus-candidate-window, I receive the error 
`E0000 00:00:1693094425.842190   36735 text_format.cc:383] Error parsing text-format mozc.ibus.Config: 10:15: Message type "mozc.ibus.Config" has no field named "mozc_renderer".`

My ibus_config.textproto looks like this when I receive the error
```
engines {
  name : "mozc-jp"
  longname : "Mozc"
  layout : "default"
  layout_variant : ""
  layout_option : ""
  rank : 80
  symbol : "あ"
}
engines {
  name : "mozc-on"
  longname : "Mozc:&#x3042"
  layout : "default"
  layout_variant : ""
  layout_option : ""
  rank : 99
  symbol : "あ"
  composition_mode : HIRAGANA
}
engines {
  name : "mozc-off"
  longname : "Mozc:A_"
  layout : "default"
  layout_variant : ""
  layout_option : ""
  rank : 99
  symbol : "A"
  composition_mode : DIRECT
}
mozc_renderer {
  # This means IBus' candidate window is always used.
  enabled : False
}
active_on_launch: True
```

**Reply by pancaek** [reply]
This persists after updating the PKGBUILD to use the latest commit from git

**Reply by yukawa** [reply]
> This persists after updating the PKGBUILD to use the latest commit from git

Hmm. That's strange. Can you restart your machine to see if the issue persists?

**Reply by yukawa** [reply]
BTW, `MOZC_IBUS_CANDIDATE_WINDOW` environment variable described in the previous revision of the document should still work.

https://github.com/google/mozc/blob/89c70080d0102e8ed23cae6c05b535dedf506de4/docs/configurations.md#use-the-default-ibus-candidate-window

**Reply by pancaek** [reply]
Strangely, restarting did not fix the issue. I used the old method, and that does seem to be working.

**Reply by yukawa** [reply]
> Strangely, restarting did not fix the issue. I used the old method, and that does seem to be working.

Then it sounds like you are still using an old build that doesn't have this commit: 65a5d1f7bdbaec800f910dcabad40a00e4031600. In my local testing environment (Ubuntu 22.04), `mozc_renderer` setting is working as intended. My suggestion is just relying on `MOZC_IBUS_CANDIDATE_WINDOW` for now and trying`mozc_renderer` out when you receive a new package release in your distribution.

**Comment by cstyan** [comment]
Should setting `mozc_renderer: False` work on sway/wlroots?

```
~ ➜ ibus version
IBus 1.5.32
~ ➜ echo $XDG_CURRENT_DESKTOP
sway:wlroots
~ ➜ sway --version
sway version 1.11
```
For mozc I am using version [2.31.5851.102])(github.com/google/mozc/releases/tag/2.31.5851.102).

I am unable to get the config file change or the env var change to be picked up so that the ibus candidate window is used rather than mozc's.

**Reply by pancaek** [reply]
it's not ibus_renderer: false you want, but the enabled key under that

```
mozc_renderer {
  # Set 'False' to use IBus' candidate window.
  enabled : False
  # For Wayland sessions, 'mozc_renderer' will be used if and only if any value
  # set in this field (e.g. "GNOME", "KDE") is found in $XDG_CURRENT_DESKTOP.
  # https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#recognized-keys
  compatible_wayland_desktop_names : ["GNOME"]
}
``` 

this is a section of what mine looks like, I'm not sure if you might need to adjust the compatible Wayland desktops, I just use gnome so I didn't have to mess with that part

**Reply by cstyan** [reply]
The comment in that section is not that clear, it could be interpreted as the entire block of `mozc_renderer` config is only used if the wayload desktop in use is present in `compatible_wayland_desktop_names`, or that the value of `enabled` is ignored for the same condition.

In either case, I tried various combinations of all those config options and nothing worked.

**Reply by yukawa** [reply]
> In either case, I tried various combinations of all those config options and nothing worked.

Any chance you have both `$HOME/.mozc/` and `$XDG_CONFIG_HOME/mozc/` in your environment?  For backward compatibility, `$HOME/.mozc/` has a higher priority than `$XDG_CONFIG_HOME/mozc/`.

https://github.com/google/mozc/blob/22ccfeb9fe04ea2b72da0d455e5a5423b0e8554a/docs/configurations.md?plain=1#L57-L61

Here are pointers to relevant functions if you try to use debugger or add logs to see what's actually happening.

https://github.com/google/mozc/blob/22ccfeb9fe04ea2b72da0d455e5a5423b0e8554a/src/base/system_util.cc#L282-L313
https://github.com/google/mozc/blob/22ccfeb9fe04ea2b72da0d455e5a5423b0e8554a/src/unix/ibus/ibus_config.cc#L54-L75

**Comment by cstyan** [comment]
I do not have a `.mozc` folder in `$HOME`, I do have `~/.config/home/mozc` which is where I was making edits to `ibus_config.textproto`. 

This is a fresh install as of ~2 weeks ago so there wouldn't be anything in `$HOME` leftover from older installs of a mozc package.

**Reply by yukawa** [reply]
> I do have `~/.config/home/mozc` which is where I was making edits to `ibus_config.textproto`.

Shouldn't it be `~/.config/mozc/ibus_config.textproto`? What is shown when you run `echo $XDG_CONFIG_HOME`?

**Reply by cstyan** [reply]
yes sorry, it is at ~/.config/mozc/ibus_config.textproto

```
~ ➜ ls ~/.config/mozc
boundary.db  cform.db  config1.db  ibus_config.textproto  segment.db

~ ➜ echo $XDG_CONFIG_HOME
/home/callum/.config
```

**Reply by yukawa** [reply]
OK, then scenarios I can imagine are:

 * A. What you are observing is already an IBus' candidate window.
   * Can you also paste the screenshot of the candidate window?
   * Check if `mozc_renderer` process is running or not. If it's already running, kill it first and try to use `ibus_mozc` again. If `mozc_renderer` starts again, then yes, `ibus_mozc` is likely to be launching it.
 * B. You forgot to restart `ibus_mozc` after updating `ibus_config.textproto`
   * Run `ibus write-cache; ibus restart`

If you are still seeing `mozc_renderer` process running after making sure that the above two are not the case, a last resort solution is just removing `mozc_renderer` file from your system.

`ibus_mozc` does check whether `mozc_renderer` exists or not, and if it does not exist always fall back to IBus' candidate window.
https://github.com/google/mozc/blob/22ccfeb9fe04ea2b72da0d455e5a5423b0e8554a/src/unix/ibus/mozc_engine.cc#L216-L221
